Close #356 - Add whitespace token for preprocessor

hdamron17 opened the pull request:
When complete, this will close #356 and maybe #395.

So far, the whitespace token has been added and properly outputs when using `-E`, but I still need to consume the whitespace before it's used by the parser.

jyn514 commented:
I'd be interested if it helped with #361 as well.

hdamron17 commented:
Observation: Need to add whitespace consideration back to preprocessor (e.g. `#ifndef x` fails because of the whitespace token in the middle) but ensure whitespace tokens have no newlines. Unit tests also fail a lot because the comparisons do not take whitespace into account.

hdamron17 commented on src/lex/mod.rs:
```diff
@@ -211,6 +214,7 @@ impl Lexer {
                 break;
```
That's what I would have thought too, but `clang -E` does not preserve comments. Not sure about `gcc` since I'm on a Mac.

jyn514 replied:
Hmm apparently `gcc -E` also strips comments. Ok, I guess this is fine then.

hdamron17 commented on src/main.rs:
```diff
@@ -134,7 +134,7 @@ fn real_main(buf: Rc<str>, bin_opt: BinOpt, output: &Path) -> Result<(), (Error,
-            write!(stdout_buf, "{} ", token.data).expect("failed to write to stdout");
+            write!(stdout_buf, "{}", token.data).expect("failed to write to stdout");
         }
         writeln!(stdout_buf).expect("failed to write to stdout");
```
Presumably the original C file ended with a newline so this would maintain the ending newline. What does this change do?

jyn514 replied:
Before the output for `-E` looked like this:
```
int main() {
	return 1;
}

```
Now it will look like this:
```
int main() {
	return 1;
}
```

jyn514 replied:
In other words, originally you would print both the original ending newline and add a newline of your own.

jyn514 commented:
Note this will not fix https://github.com/jyn514/rcc/issues/395. The `\n` in the issue title refers to the actual characters `\` and then `n`, not a newline.

hdamron17 commented:
It's working now, but I probably should put some whitespace-dependent test cases for `-E`.

One possible issue that came up is that the parser iteration does not work if there is leading whitespace. However, it works in the final product so I don't think it matters. I changed the test case to have no leading whitespace in 8e0509e.

hdamron17 commented:
Yeah it doesn't keep whitespace properly, mostly with preprocessor stuff since I was just trying to get existing tests to pass...

jyn514 commented:
> One possible issue that came up is that the parser iteration does not work if there is leading whitespace. However, it works in the final product so I don't think it matters. I changed the test case to have no leading whitespace in 8e0509e.

If I understand right, this means that it works correctly using `check_semantics` but that passing a whitespace token to `Parser` as `first` will give a spurious error? That seems fine as long as it's documented.

jyn514 commented on src/lex/cpp.rs:
```diff
@@ -886,6 +886,15 @@ impl<'a> PreProcessor<'a> {
+                        vec![x, Ok(Token::Whitespace(String::from(" ")))]
+                    }
+                })
+                .flatten()
```
Why do you add these spaces here in the define? Does `tokens_until_newline` not return whitespace tokens?

hdamron17 replied:
`tokens_until_newline` (at the moment) does not include whitespace tokens. I did not want to change it since it is somewhat separate from the preprocessor. I can try adding whitespace tokens to it and see what happens now that it is in a stable state. I did notice that `clang` only puts one space when replacing preprocessor defines, regardless of the original spacing.

jyn514 replied:
Hmm interesting ... I suppose since the behavior is correct we can try to go back and improve the spacing later.

jyn514 replied:
Also, please add a comment to this effect either here or at `tokens_until_newline`.

hdamron17 replied:
I actually went back and did it the proper way by changing tokens_until_newline. Unfortunately I had to rework some stuff for boolean_expr because whitespace show up in the replacement stage. I'll push as soon as I double check the tests. Also, I think the rework will be much less of an eyesore.

jyn514 commented on src/lex/mod.rs:
```diff
@@ -191,22 +191,29 @@ impl Lexer {
             while self.peek().map_or(false, |c| c.is_ascii_whitespace()) {
-                self.next_char();
+                if let Some(c) = self.next_char() {
+                    whitespace.push(c.into());
```
This might now be better as a match + loop so you don't have to check the `Some` condition twice. Up to you though.

hdamron17 replied:
Not sure what you mean by that, but this doesn't look too bad to me.

jyn514 commented on src/lex/mod.rs:
```diff
@@ -865,7 +905,8 @@ impl Iterator for Lexer {
             };
-            self.seen_line_token |= data != Token::Hash;
+            // We've seen a token if this isn't # and this isn't whitespace
+            self.seen_line_token |= data != Token::Hash && !matches!(data, Token::Whitespace(_));
```
This might be easier to read if you reversed the condition: `!(Token::Hash || whitespace)`.

hdamron17 replied:
Done

hdamron17 commented:
Some more tests with preprocessor stuff would be nice, but other than that, I think it's complete.

hdamron17 commented:
Also, newlines are not preserved for preprocessor macros at the moment, so I'll need to fix that.

jyn514 commented:
> Also, newlines are not preserved for preprocessor macros at the moment, so I'll need to fix that.

Do you mean that `\\\n` isn't preserved?
```c
#define f(a) { \
	int b = a; \
	return b; \
}
f(1)
 { int b = 1; return b; }
```
If so, that's fine, clang does the same. It will also be hard to do without major changes since deleting `\\\n` happens very early in the lexer: https://github.com/jyn514/rcc/blob/93b5e06/src/lex/mod.rs#L98

jyn514 commented on src/lex/cpp.rs:
```diff
@@ -1215,19 +1256,34 @@ mod tests {
+    fn assert_same_exact(src: &str, cpp_src: &str) {
+        // NOTE make sure `cpp_src` has a trailing newline
+        let pprint = cpp(src)
+            .filter_map(|res| res.ok().map(|token| token.data))
+            .map(|res| format!("{}", res))
```
```suggestion
            .filter_map(|res| res.ok().map(|token| token.data.to_string()))
```

jyn514 commented on src/lex/cpp.rs:
```diff
@@ -1534,4 +1589,12 @@ int main(){}
+    fn preprocess_only() {
+        assert_same_exact("int \t\n\r     main() {}", "int \t\n\r     main() {}");
+        assert_same_exact("int/* */main() {}", "int main() {}");
+        assert_same_exact("int/*\n\n\n*/main() {}", "int\n\n\nmain() {}");
```
Hmm, this behavior looks a little confusing. Is there a reason you kept newlines inside of block comments?

hdamron17 replied:
This seems to be the behavior of clang. For example,
```
int main() {
 /*




 */
}
```
preprocesses to
```
# 1 "test.c"
# 1 "<built-in>" 1
# 1 "<built-in>" 3
# 363 "<built-in>" 3
# 1 "<command line>" 1
# 1 "<built-in>" 2
# 1 "test.c" 2
int main() {






}
```

hdamron17 replied:
I may need to look at the documentation though since
```
int main() {/*




*/}
```
mysteriously preprocesses to
```
# 1 "test.c"
# 1 "<built-in>" 1
# 1 "<built-in>" 3
# 363 "<built-in>" 3
# 1 "<command line>" 1
# 1 "<built-in>" 2
# 1 "test.c" 2
int main() { }
```

jyn514 replied:
Relevant part of the standard ([5.1.1.2 Translation phases](http://port70.net/~nsz/c/c11/n1570.html#5.1.1.2)):
>  The source file is decomposed into preprocessing tokens7) and sequences of white-space characters (including comments). A source file shall not end in a partial preprocessing token or in a partial comment. Each comment is replaced by one space character. New-line characters are retained. Whether each nonempty sequence of white-space characters other than new-line is retained or replaced by one space character is implementation-defined.

This seems to me to say that clang's behavior in your first example is correct, and the behavior in the second is a bug. For context, gcc always behaves like clang in the second example, i.e. it always deletes newlines in comments:

```c
gcc -x c -E -P -
int main() {
 /*




 */
}
```
preprocesses to
```c
int main() {
}
```
and
```c
int main() {/*




*/}
```
preprocesses to `int main() { }`.

jyn514 replied:
tcc behaves the same as gcc.

hdamron17 commented:
> > One possible issue that came up is that the parser iteration does not work if there is leading whitespace. However, it works in the final product so I don't think it matters. I changed the test case to have no leading whitespace in [8e0509e](https://github.com/jyn514/rcc/commit/8e0509e45ef518c0c3b2583b92de0cc99aa0e096).
> 
> If I understand right, this means that it works correctly using `check_semantics` but that passing a whitespace token to `Parser` as `first` will give a spurious error? That seems fine as long as it's documented.

Turns out it was an easy fix. I just changed the `parser` function to use `next_non_whitespace`.

hdamron17 commented:
> > Also, newlines are not preserved for preprocessor macros at the moment, so I'll need to fix that.
> 
> Do you mean that `\\\n` isn't preserved?

No, I mean `#define`s do not keep newlines after.
```
int main() {
#define a
#define b
#define c
}
```
preprocesses to
```
int main() {
}
```
but should preprocess to
```
# 1 "test.c"
# 1 "<built-in>" 1
# 1 "<built-in>" 3
# 363 "<built-in>" 3
# 1 "<command line>" 1
# 1 "<built-in>" 2
# 1 "test.c" 2
int main() {



}
```

hdamron17 commented:
The newlines not being preserved after some macros is because `tokens_until_newline` does not actually stop at the newline but just consumes arbitrary whitespace until the line number increases. For example,
```
int main() {
#define a 1
           return a;
}
```
preprocesses to
```
int main() {
return 1;
}
```

jyn514 commented:
> tokens_until_newline does not actually stop at the newline but just consumes arbitrary whitespace until the line number increases

I don't see a way to fix this without fully separating the preprocessor from the lexer ... I guess I could make that function part of the lexer instead and stop at `\n`? That would require a fair bit of refactoring but not an unreasonable amount.

jyn514 commented:
In any case I think it can be fixed later, the majority of things are working now.

jyn514 commented:
r=me once @hdamron17 is happy with it

hdamron17 commented:
I went ahead and changed `tokens_until_newline` and it helps with `#defines`. All that's missing now are `#else` and `#endif` I think so I'll try to get those working and make enough tests to be worthwhile.

jyn514 commented on src/lex/cpp.rs:
```diff
@@ -1594,6 +1595,8 @@ int main(){}
-        // TODO add tests for `assert_same_exact` with preprocessor macros
+        assert_same_exact("#define a(c) c\tc\na(1);a(2)", "\n1\t1;2\t2");
+        // assert_same_exact("#define a //\n#if defined a\n  x\n#endif", "\n\n  x");
+        // TODO add mote tests for `assert_same_exact` with preprocessor macros
```
```suggestion
        // TODO add more tests for `assert_same_exact` with preprocessor macros
```

jyn514 commented on src/lex/cpp.rs:
```diff
@@ -880,17 +894,14 @@ impl<'a> PreProcessor<'a> {
-            this.tokens_until_newline()
+            this.tokens_until_newline(true)
                 .into_iter()
+                .skip(1)  // First is always unwanted whitespace
```
Are you sure this is true? What about things like `#define f+`? This has different behavior on your branch than on master (and probably deserves an automated test).

jyn514 replied:
Ah interesting, clang gives an error if there's not a space. We should do the same.

```
$ clang -x c - -E -pedantic-errors -P
#define f+
f
<stdin>:1:10: error: ISO C99 requires whitespace after the macro name [-Werror,-Wc99-extensions]
```

hdamron17 replied:
"warning: ISO C99 requires whitespace after the macro name"
I guess I should at least error here if it's not whitespace instead of blindly assuming it is.

hdamron17 replied:
Whitespace is no longer required between `#define` id and body. I do not give a warning if there is no whitespace, though.

jyn514 commented on src/lex/cpp.rs:
```diff
@@ -1215,19 +1257,33 @@ mod tests {
             _ => panic!("not a keyword: {:?}", token),
         }
     }
+    fn _is_not_whitespace(res: &CompileResult<Token>) -> bool {
+        !matches!(res, Ok(Token::Whitespace(_)))
+    }
```
Why is there both this function and `is_not_whitespace`?

hdamron17 replied:
I was being dumb and applied the second one in a different order so I though I needed two with different signatures Changed in coming commit.

hdamron17 commented:
r @jyn514

hdamron17 commented:
Your summary sounds about right. Also the behaviour of `consume_whitespace_oneline` was just copied from your code so I guess you're the clever one. :)
